// office-viewer/esm/package/ZipPackageParser.js
import { unzipSync, strFromU8, strToU8, zipSync } from 'fflate';

/**
 * zip 文件解析
 */
var ZipPackageParser = /** @class */ (function () {
    function ZipPackageParser() {
    }
    /**
     * 加载 zip 文件
     */
    ZipPackageParser.prototype.load = function (docxFile) {
        this.zip = unzipSync(new Uint8Array(docxFile));
    };
    /**
     * 读取 xml 文件，转成 json 对象
     * @param filePath 文件路径
     * @returns 转成 json 的结果
     */
    ZipPackageParser.prototype.getXML = function (filePath) {
        var fileContent = this.getFileByType(filePath, 'string');
        var doc = new DOMParser().parseFromString(fileContent, 'application/xml');
        var errorNode = doc.getElementsByTagName('parsererror').item(0);
        if (errorNode) {
            throw new Error(errorNode.textContent || "can't parse xml");
        }
        else {
            return doc;
        }
    };
    /**
     * 根据类型读取文件
     */
    ZipPackageParser.prototype.getFileByType = function (filePath, type) {
        filePath = filePath.startsWith('/') ? filePath.slice(1) : filePath;
        var file = this.zip[filePath];
        if (file) {
            if (type === 'string') {
                return strFromU8(file);
            }
            else if (type === 'blob') {
                return new Blob([file]);
            }
            else if (type === 'uint8array') {
                return file;
            }
        }
        console.warn('getFileByType', filePath, 'not found');
        return null;
    };
    /**
     * xml 下没这功能
     */
    ZipPackageParser.prototype.saveFile = function (filePath, content) {
        if (typeof content === 'string') {
            content = strToU8(content);
        }
        this.zip[filePath] = content;
    };
    /**
     * 判断文件是否存在
     */
    ZipPackageParser.prototype.fileExists = function (filePath) {
        filePath = filePath.startsWith('/') ? filePath.slice(1) : filePath;
        return filePath in this.zip;
    };
    /**
     * 生成新的 zip 文件
     */
    ZipPackageParser.prototype.generateZip = function (docContent) {
        // 其实最好是生成个新的，后续再优化
        this.zip['word/document.xml'] = strToU8(docContent);
        return new Blob([zipSync(this.zip)]);
    };
    return ZipPackageParser;
}());

export { ZipPackageParser as default };
