// amis-ui/esm/components/formula/Picker.js
/**
 * amis-ui v3.6.3
 * Copyright 2018-2023 fex
 */

import { __extends, __awaiter, __generator, __assign, __rest, __read, __decorate, __metadata } from 'tslib';
import { parse, noop, autobind, themeable, localeable, uncontrollable, isExpression, resolveVariableAndFilterForAsync } from 'amis-core';
import React__default from 'react';
import Editor, { FormulaEditor } from './Editor.js';
import ResultBox from '../ResultBox.js';
import Button from '../Button.js';
import { Icon } from '../icons.js';
import FinalModal from '../Modal.js';
import PopUp from '../PopUp.js';
import FormulaInput from './Input.js';

var InputSchemaType = [
    'text',
    'number',
    'boolean',
    'date',
    'time',
    'datetime',
    'select'
];
var FormulaPicker = /** @class */ (function (_super) {
    __extends(FormulaPicker, _super);
    function FormulaPicker(props) {
        var _this = _super.call(this, props) || this;
        _this.props.onRef && _this.props.onRef(_this);
        _this.state = {
            isOpened: false,
            value: _this.props.value,
            editorValue: _this.value2EditorValue(_this.props),
            isError: false,
            variables: Array.isArray(props.variables) ? props.variables : []
        };
        return _this;
    }
    FormulaPicker.prototype.componentDidUpdate = function (prevProps) {
        var value = this.props.value;
        if (value !== prevProps.value) {
            this.setState({
                value: typeof value === 'string' || !this.isTextInput() ? value : '',
                editorValue: this.value2EditorValue(this.props)
            });
        }
    };
    FormulaPicker.prototype.value2EditorValue = function (props) {
        var value = props.value;
        if (!this.isTextInput()) {
            var editorValue = '';
            try {
                editorValue = JSON.stringify(value);
            }
            catch (error) { }
            return editorValue;
        }
        if (props.mixedMode) {
            if (typeof props.value === 'string' &&
                /^\s*\$\{(.+?)\}\s*$/.test(props.value)) {
                return RegExp.$1;
            }
            else {
                return '';
            }
        }
        return String(props.value || '');
    };
    FormulaPicker.prototype.isTextInput = function () {
        var inputSettings = this.props.inputSettings;
        return (!inputSettings ||
            (inputSettings === null || inputSettings === void 0 ? void 0 : inputSettings.type) === 'text' ||
            !InputSchemaType.includes(inputSettings === null || inputSettings === void 0 ? void 0 : inputSettings.type));
    };
    FormulaPicker.prototype.handleConfirm = function () {
        var _a, _b;
        var value = this.state.value;
        if (this.props.onConfirm) {
            this.props.onConfirm(value);
        }
        else {
            (_b = (_a = this.props).onChange) === null || _b === void 0 ? void 0 : _b.call(_a, value);
        }
    };
    FormulaPicker.prototype.renderFormulaValue = function (item) {
        var _a = this.props, allowInput = _a.allowInput, cx = _a.classnames;
        var html = { __html: item.html };
        if (allowInput) {
            return '';
        }
        return (React__default.createElement("div", { className: cx('FormulaPicker-ResultBox'), dangerouslySetInnerHTML: html }));
    };
    FormulaPicker.prototype.handleInputChange = function (value) {
        var _this = this;
        this.setState({ value: value }, function () { return _this.handleConfirm(); });
    };
    FormulaPicker.prototype.handleInputGroupChange = function (e) {
        var onChange = this.props.onChange;
        onChange && onChange(e.currentTarget.value);
    };
    FormulaPicker.prototype.handleEditorChange = function (value) {
        this.setState({
            editorValue: value,
            isError: false
        });
    };
    FormulaPicker.prototype.handleEditorConfirm = function () {
        var _a, _b;
        var _c = this.props; _c.translate; var inputSettings = _c.inputSettings;
        var editorValue = this.state.editorValue;
        if (this.isTextInput()) {
            return this.confirm(editorValue);
        }
        else if (inputSettings) {
            var result = editorValue;
            var schemaType = inputSettings === null || inputSettings === void 0 ? void 0 : inputSettings.type;
            try {
                var ast = parse(editorValue, { evalMode: true, allowFilter: false });
                if (schemaType === 'select' &&
                    inputSettings.multiple &&
                    ast.type === 'array') {
                    result = ast.members.map(function (i) { return i.value; });
                }
                else if (ast.type === 'literal' || ast.type === 'string') {
                    result = (_a = ast.value) !== null && _a !== void 0 ? _a : '';
                }
            }
            catch (error) {
                this.setState({ isError: (_b = error === null || error === void 0 ? void 0 : error.message) !== null && _b !== void 0 ? _b : true });
                return;
            }
            this.setState({ isError: false });
            return this.confirm(result);
        }
    };
    FormulaPicker.prototype.confirm = function (value) {
        var _this = this;
        var mixedMode = this.props.mixedMode;
        var validate = this.validate(value);
        if (validate === true) {
            this.setState({ value: mixedMode && value ? "${".concat(value, "}") : value }, function () {
                _this.close(undefined, function () { return _this.handleConfirm(); });
            });
        }
        else {
            this.setState({ isError: validate });
        }
    };
    FormulaPicker.prototype.handleClick = function () {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            var _c, variables, data, list, result, state, _d;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        _c = this.props, variables = _c.variables, data = _c.data;
                        if (!(typeof variables === 'function')) return [3 /*break*/, 2];
                        return [4 /*yield*/, variables(this.props)];
                    case 1:
                        list = _e.sent();
                        this.setState({ variables: list });
                        return [3 /*break*/, 4];
                    case 2:
                        if (!(typeof variables === 'string' && isExpression(variables))) return [3 /*break*/, 4];
                        return [4 /*yield*/, resolveVariableAndFilterForAsync(variables, data, '|raw')];
                    case 3:
                        result = _e.sent();
                        this.setState({ variables: result });
                        _e.label = 4;
                    case 4:
                        _d = [{}];
                        return [4 /*yield*/, ((_b = (_a = this.props).onPickerOpen) === null || _b === void 0 ? void 0 : _b.call(_a, this.props))];
                    case 5:
                        state = __assign.apply(void 0, [__assign.apply(void 0, _d.concat([(_e.sent())])), { editorValue: this.value2EditorValue(this.props), isOpened: true }]);
                        this.setState(state);
                        return [2 /*return*/];
                }
            });
        });
    };
    FormulaPicker.prototype.close = function (e, callback) {
        this.setState({
            isOpened: false,
            isError: false
        }, function () {
            if (callback) {
                callback();
                return;
            }
        });
    };
    FormulaPicker.prototype.updateState = function (state) {
        if (state === void 0) { state = {}; }
        state.isOpened; var rest = __rest(state, ["isOpened"]);
        this.setState(__assign(__assign({}, this.state), rest));
    };
    FormulaPicker.prototype.validate = function (value) {
        var _a = this.props, __ = _a.translate, inputSettings = _a.inputSettings;
        /** 处理非文本输入场景 */
        if (inputSettings && !this.isTextInput()) {
            var schemaType = inputSettings === null || inputSettings === void 0 ? void 0 : inputSettings.type;
            var errorMsg = __('FormulaEditor.invalidValue');
            /** 变量类型 */
            if (typeof value === 'string') {
                return true;
            }
            if (['number', 'boolean'].includes(schemaType)) {
                return typeof value === schemaType ? true : errorMsg;
            }
            else if (['text', 'date', 'time', 'datetime'].includes(schemaType)) {
                return typeof value === 'string' ? true : errorMsg;
            }
            else if (schemaType === 'select' && inputSettings.multiple) {
                return Array.isArray(value) ? true : errorMsg;
            }
            else {
                return true;
            }
        }
        try {
            value &&
                parse(value, {
                    evalMode: this.props.mixedMode ? true : this.props.evalMode,
                    allowFilter: false
                });
            return true;
        }
        catch (e) {
            if (/\s(\d+:\d+)$/.test(e.message)) {
                var _b = __read(/\s(\d+:\d+)$/.exec(e.message) || [], 2), position = _b[1];
                return position;
            }
            return e.message;
        }
    };
    FormulaPicker.prototype.render = function () {
        var _a, _b, _c;
        var _d, _e, _f, _g, _h, _j;
        var _k = this.props, cx = _k.classnames, __ = _k.translate, disabled = _k.disabled, _l = _k.allowInput, allowInput = _l === void 0 ? true : _l, className = _k.className, style = _k.style; _k.onChange; _k.size; var borderMode = _k.borderMode, placeholder = _k.placeholder, _m = _k.mode, mode = _m === void 0 ? 'input-button' : _m, btnLabel = _k.btnLabel, level = _k.level, btnSize = _k.btnSize, icon = _k.icon, title = _k.title, clearable = _k.clearable, functions = _k.functions, children = _k.children, variableMode = _k.variableMode, mixedMode = _k.mixedMode, evalMode = _k.evalMode, popOverContainer = _k.popOverContainer, mobileUI = _k.mobileUI, inputSettings = _k.inputSettings, rest = __rest(_k, ["classnames", "translate", "disabled", "allowInput", "className", "style", "onChange", "size", "borderMode", "placeholder", "mode", "btnLabel", "level", "btnSize", "icon", "title", "clearable", "functions", "children", "variableMode", "mixedMode", "evalMode", "popOverContainer", "mobileUI", "inputSettings"]);
        var _o = this.state, isOpened = _o.isOpened, value = _o.value, editorValue = _o.editorValue, isError = _o.isError;
        var iconElement = React__default.createElement(Icon, { cx: cx, icon: icon, className: "Icon" });
        return (React__default.createElement(React__default.Fragment, null,
            children ? (children({
                isOpened: this.state.isOpened,
                onClick: this.handleClick,
                setState: this.updateState
            })) : (React__default.createElement("div", { className: cx('FormulaPicker', mode === 'input-group' ? 'is-input-group' : '', {
                    'FormulaPicker--text': this.isTextInput()
                }, className), style: style },
                mode === 'button' && (React__default.createElement(Button, { className: cx('FormulaPicker-action', 'w-full'), level: level, size: btnSize, onClick: this.handleClick },
                    iconElement ? (React__default.cloneElement(iconElement, {
                        className: cx((_e = (_d = iconElement === null || iconElement === void 0 ? void 0 : iconElement.props) === null || _d === void 0 ? void 0 : _d.className) !== null && _e !== void 0 ? _e : '', 'FormulaPicker-icon', (_a = {},
                            _a['is-filled'] = !!value,
                            _a))
                    })) : (React__default.createElement(Icon, { icon: "function", className: cx('FormulaPicker-icon', 'icon', (_b = {},
                            _b['is-filled'] = !!value,
                            _b)) })),
                    React__default.createElement("span", { className: cx('FormulaPicker-label') }, __(btnLabel || 'FormulaEditor.btnLabel')))),
                mode === 'input-button' && (React__default.createElement(React__default.Fragment, null,
                    React__default.createElement(ResultBox, { className: cx('FormulaPicker-input', isOpened ? 'is-active' : '', !!isError ? 'is-error' : ''), allowInput: allowInput, clearable: clearable, value: value, result: allowInput
                            ? void 0
                            : FormulaEditor.highlightValue(value, this.state.variables, this.props.evalMode), itemRender: this.renderFormulaValue, onResultChange: noop, onChange: this.handleInputChange, disabled: disabled, borderMode: borderMode, placeholder: placeholder }),
                    React__default.createElement(Button, { className: cx('FormulaPicker-action'), onClick: this.handleClick },
                        React__default.createElement(Icon, { icon: "function", className: cx('FormulaPicker-icon', 'icon', (_c = {},
                                _c['is-filled'] = !!value,
                                _c)) })))),
                mode === 'input-group' && (React__default.createElement(React__default.Fragment, null,
                    React__default.createElement(FormulaInput, { className: cx('FormulaPicker-input', isOpened ? 'is-active' : '', !!isError ? 'is-error' : ''), inputSettings: inputSettings, allowInput: allowInput, clearable: clearable, evalMode: evalMode, mixedMode: mixedMode, variables: this.state.variables, value: value, itemRender: this.renderFormulaValue, onChange: this.handleInputChange, disabled: disabled, borderMode: borderMode, placeholder: placeholder }),
                    React__default.createElement("a", { className: cx("FormulaPicker-toggler"), onClick: this.handleClick },
                        React__default.createElement(Icon, { icon: "function", className: "icon" })))))),
            mobileUI ? (React__default.createElement(PopUp, { className: cx("FormulaPicker-popup"), isShow: this.state.isOpened, showConfirm: true, onHide: this.close, onConfirm: this.handleEditorConfirm, container: popOverContainer },
                React__default.createElement("div", { className: cx('FormulaPicker-popup-inner') },
                    React__default.createElement(Editor, __assign({}, rest, { evalMode: mixedMode ? true : evalMode, variables: this.state.variables, functions: (_f = this.state.functions) !== null && _f !== void 0 ? _f : functions, variableMode: (_g = this.state.variableMode) !== null && _g !== void 0 ? _g : variableMode, value: editorValue, onChange: this.handleEditorChange, selfVariableName: this.props.selfVariableName })),
                    !!isError ? (React__default.createElement("div", { className: cx('Dialog-info'), key: "info" },
                        React__default.createElement("span", { className: cx('Dialog-error') }, __('FormulaEditor.invalidData', { err: isError })))) : null))) : (React__default.createElement(FinalModal, { size: "lg", closeOnEsc: true, show: this.state.isOpened, onHide: this.close, container: popOverContainer },
                React__default.createElement(FinalModal.Header, { onClose: this.close, className: "font-bold" }, __(title || 'FormulaEditor.title')),
                React__default.createElement(FinalModal.Body, null,
                    React__default.createElement(Editor, __assign({}, rest, { evalMode: mixedMode ? true : evalMode, variables: this.state.variables, functions: (_h = this.state.functions) !== null && _h !== void 0 ? _h : functions, variableMode: (_j = this.state.variableMode) !== null && _j !== void 0 ? _j : variableMode, value: editorValue, onChange: this.handleEditorChange, selfVariableName: this.props.selfVariableName }))),
                React__default.createElement(FinalModal.Footer, null,
                    !!isError ? (React__default.createElement("div", { className: cx('Dialog-info'), key: "info" },
                        React__default.createElement("span", { className: cx('Dialog-error') }, __('FormulaEditor.invalidData', { err: isError })))) : null,
                    React__default.createElement(Button, { onClick: this.close }, __('cancel')),
                    React__default.createElement(Button, { onClick: this.handleEditorConfirm, level: "primary" }, __('confirm')))))));
    };
    FormulaPicker.defaultProps = {
        evalMode: true
    };
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], FormulaPicker.prototype, "handleConfirm", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], FormulaPicker.prototype, "renderFormulaValue", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], FormulaPicker.prototype, "handleInputChange", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], FormulaPicker.prototype, "handleInputGroupChange", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], FormulaPicker.prototype, "handleEditorChange", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], FormulaPicker.prototype, "handleEditorConfirm", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], FormulaPicker.prototype, "handleClick", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object, Function]),
        __metadata("design:returntype", void 0)
    ], FormulaPicker.prototype, "close", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], FormulaPicker.prototype, "updateState", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], FormulaPicker.prototype, "validate", null);
    return FormulaPicker;
}(React__default.Component));
var FormulaPicker$1 = themeable(localeable(uncontrollable(FormulaPicker, {
    value: 'onChange'
})));

export { FormulaPicker, InputSchemaType, FormulaPicker$1 as default };
