// amis-core/esm/utils/api.js
/**
 * amis-core v3.6.3
 * Copyright 2018-2023 fex
 */

import { __awaiter, __generator, __assign, __rest, __spreadArray, __read } from 'tslib';
import omit from 'lodash/omit';
import { evaluate } from 'amis-formula';
import { escapeHtml } from './escapeHtml.js';
import 'moment';
import { cloneObject, extendObject, createObject, isObject } from './object.js';
import { tokenize } from './tokenize.js';
import { memoryParse } from './memoryParse.js';
import { dataMapping } from './dataMapping.js';
import './filter.js';
import { evalExpression } from './tpl.js';
import { qsparse, JSONTraverse, isEmpty, uuid, qsstringify, isObjectShallowModified, hasFile, object2formData } from './helper.js';
import isPlainObject from 'lodash/isPlainObject';
import { debug } from './debug.js';

var rSchema = /(?:^|raw\:)(get|post|put|delete|patch|options|head|jsonp|js):/i;
var apiCaches = [];
var isIE = !!document.documentMode;
function normalizeApi(api, defaultMethod) {
    if (defaultMethod === void 0) { defaultMethod = 'get'; }
    if (typeof api === 'string') {
        var method = rSchema.test(api) ? RegExp.$1 : '';
        method && (api = api.replace(method + ':', ''));
        api = {
            method: (method || defaultMethod),
            url: api
        };
    }
    else {
        api = __assign({}, api);
    }
    api.url = typeof api.url === 'string' ? api.url.trim() : api.url;
    return api;
}
function buildApi(api, data, options) {
    if (options === void 0) { options = {}; }
    api = normalizeApi(api, options.method);
    var autoAppend = options.autoAppend, ignoreData = options.ignoreData, rest = __rest(options, ["autoAppend", "ignoreData"]);
    api.config = __assign({}, rest);
    api.method = (api.method || options.method || 'get').toLowerCase();
    if (api.headers) {
        api.headers = dataMapping(api.headers, data, undefined, false);
    }
    if (api.requestAdaptor && typeof api.requestAdaptor === 'string') {
        api.requestAdaptor = str2AsyncFunction(api.requestAdaptor, 'api', 'context');
    }
    if (api.adaptor && typeof api.adaptor === 'string') {
        api.adaptor = str2AsyncFunction(api.adaptor, 'payload', 'response', 'api', 'context');
    }
    if (!data) {
        return api;
    }
    else if (data instanceof FormData ||
        data instanceof Blob ||
        data instanceof ArrayBuffer) {
        api.data = data;
        return api;
    }
    var raw = (api.url = api.url || '');
    var ast = undefined;
    try {
        ast = memoryParse(api.url);
    }
    catch (e) {
        console.warn("api \u914D\u7F6E\u8BED\u6CD5\u51FA\u9519\uFF1A".concat(e));
        return api;
    }
    var url = ast.body
        .map(function (item, index) {
        return item.type === 'raw' ? item.value : "__expression__".concat(index, "__");
    })
        .join('');
    var idx = url.indexOf('?');
    var replaceExpression = function (fragment, defaultFilter, defVal) {
        if (defaultFilter === void 0) { defaultFilter = 'url_encode'; }
        if (defVal === void 0) { defVal = undefined; }
        return fragment.replace(/__expression__(\d+)__/g, function (_, index) {
            var _a;
            return ((_a = evaluate(ast.body[index], data, {
                defaultFilter: defaultFilter
            })) !== null && _a !== void 0 ? _a : defVal);
        });
    };
    // 是否过滤空字符串 query
    var queryStringify = function (query) {
        return qsstringify(query, (api === null || api === void 0 ? void 0 : api.filterEmptyQuery)
            ? {
                filter: function (key, value) {
                    return value === '' ? undefined : value;
                }
            }
            : undefined);
    };
    /** 追加data到请求的Query中 */
    var attachDataToQuery = function (apiObject, ctx, merge) {
        var idx = apiObject.url.indexOf('?');
        if (~idx) {
            var params = (apiObject.query = __assign(__assign(__assign({}, qsparse(apiObject.url.substring(idx + 1))), apiObject.query), ctx));
            apiObject.url =
                apiObject.url.substring(0, idx) + '?' + queryStringify(params);
        }
        else {
            apiObject.query = __assign(__assign({}, apiObject.query), ctx);
            var query = queryStringify(merge ? apiObject.query : ctx);
            if (query) {
                apiObject.url = "".concat(apiObject.url, "?").concat(query);
            }
        }
        return apiObject;
    };
    if (~idx) {
        var hashIdx = url.indexOf('#');
        var params = qsparse(url.substring(idx + 1, ~hashIdx && hashIdx > idx ? hashIdx : undefined));
        // 将里面的表达式运算完
        JSONTraverse(params, function (value, key, host) {
            var _a;
            if (typeof value === 'string' && /^__expression__(\d+)__$/.test(value)) {
                host[key] = (_a = evaluate(ast.body[RegExp.$1].body, data)) !== null && _a !== void 0 ? _a : '';
            }
            else if (typeof value === 'string') {
                // 参数值里面的片段不能 url_encode 了，所以是不处理
                host[key] = replaceExpression(host[key], 'raw', '');
            }
        });
        var left = replaceExpression(url.substring(0, idx), 'raw', '');
        // 追加
        Object.assign(params, api.query);
        api.url =
            left +
                (~left.indexOf('?') ? '&' : '?') +
                queryStringify((api.query = dataMapping(params, data, undefined, api.convertKeyToPath))) +
                (~hashIdx && hashIdx > idx
                    ? replaceExpression(url.substring(hashIdx))
                    : '');
    }
    else {
        api.url = replaceExpression(url, 'raw', '');
    }
    if (ignoreData) {
        return api;
    }
    var hasCustomData = api.data;
    if (api.data) {
        api.body = api.data = dataMapping(api.data, data, undefined, api.convertKeyToPath);
    }
    else if (api.method === 'post' ||
        api.method === 'put' ||
        api.method === 'patch') {
        api.body = api.data = data;
    }
    // 给 query 做数据映射
    if (api.query) {
        api.query = dataMapping(api.query, data, undefined, api.convertKeyToPath);
    }
    // get 类请求，把 data 附带到 url 上。
    if (api.method === 'get' || api.method === 'jsonp' || api.method === 'js') {
        if (!api.data &&
            ((!~raw.indexOf('$') && autoAppend) || api.forceAppendDataToQuery)) {
            api.data = data;
            api.query = __assign(__assign({}, api.query), data);
        }
        else if (api.attachDataToQuery === false &&
            api.data &&
            ((!~raw.indexOf('$') && autoAppend) || api.forceAppendDataToQuery)) {
            api = attachDataToQuery(api, data, false);
        }
        if (api.data && api.attachDataToQuery !== false) {
            api = attachDataToQuery(api, api.data, true);
            delete api.data;
        }
    }
    // 非 get 类请求也可以携带参数到 url，只要 query 有值
    else if (api.method) {
        var idx_1 = api.url.indexOf('?');
        if (~idx_1) {
            var params = (api.query = __assign(__assign({}, qsparse(api.url.substring(idx_1 + 1))), api.query));
            api.url = api.url.substring(0, idx_1) + '?' + queryStringify(params);
        }
        else {
            var query = queryStringify(api.query);
            if (query) {
                api.url = "".concat(api.url, "?").concat(query);
            }
        }
    }
    if (api.graphql) {
        if (api.method === 'get') {
            api.query = api.data = __assign(__assign({}, api.query), { query: api.graphql });
        }
        else if (api.method === 'post' ||
            api.method === 'put' ||
            api.method === 'patch') {
            api.body = api.data = {
                query: api.graphql,
                operationName: api.operationName,
                variables: cloneObject(api.data)
            };
        }
    }
    else if (api.jsonql) {
        api.method = 'post';
        api.jsonql = dataMapping(api.jsonql, 
        /** 需要上层数据域的内容 */
        extendObject(data, __assign(__assign({}, api.query), data), false), undefined, false, true);
        /** 同时设置了JSONQL和data时走兼容场景 */
        api.body = api.data =
            hasCustomData && api.jsonql
                ? {
                    data: api.data,
                    jsonql: api.jsonql
                }
                : api.jsonql;
        /** JSONQL所有method需要追加data中的变量到query中 */
        if (api.forceAppendDataToQuery) {
            api = attachDataToQuery(api, data, true);
        }
    }
    return api;
}
function str2function(contents) {
    var args = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        args[_i - 1] = arguments[_i];
    }
    try {
        var fn = new (Function.bind.apply(Function, __spreadArray(__spreadArray([void 0], __read(args), false), [contents], false)))();
        return fn;
    }
    catch (e) {
        console.warn(e);
        return null;
    }
}
var AsyncFunction = Object.getPrototypeOf(function () {
    return __awaiter(this, void 0, void 0, function () { return __generator(this, function (_a) {
        return [2 /*return*/];
    }); });
}).constructor;
function str2AsyncFunction(contents) {
    var args = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        args[_i - 1] = arguments[_i];
    }
    try {
        var fn = new (AsyncFunction.bind.apply(AsyncFunction, __spreadArray(__spreadArray([void 0], __read(args), false), [contents], false)))();
        return fn;
    }
    catch (e) {
        console.warn(e);
        return null;
    }
}
function callStrFunction(fn, argNames) {
    var args = [];
    for (var _i = 2; _i < arguments.length; _i++) {
        args[_i - 2] = arguments[_i];
    }
    if (typeof fn === 'function') {
        return fn.apply(this, args);
    }
    else if (typeof fn === 'string' && fn) {
        var func = str2function.apply(void 0, __spreadArray([fn], __read(argNames), false));
        return func === null || func === void 0 ? void 0 : func.apply(this, args);
    }
}
function responseAdaptor(ret, api) {
    var data = ret.data;
    var hasStatusField = true;
    if (!data) {
        throw new Error('Response is empty');
    }
    // 返回内容是 string，说明 content-type 不是 json，这时可能是返回了纯文本或 html
    if (typeof data === 'string') {
        var contentType = ret.headers['content-type'] || '';
        // 如果是文本类型就尝试解析一下
        if (ret.headers &&
            contentType.startsWith('text/') &&
            !contentType.includes('markdown') &&
            api.responseType !== 'blob') {
            try {
                data = JSON.parse(data);
                if (typeof data === 'undefined') {
                    throw new Error('Response should be JSON');
                }
            }
            catch (e) {
                var responseBrief = typeof data === 'string'
                    ? escapeHtml(data.substring(0, 100))
                    : '';
                throw new Error("Response should be JSON\n ".concat(responseBrief));
            }
        }
        else {
            if (api.responseType === 'blob') {
                throw new Error('Should have "Content-Disposition" in Header');
            }
            else if (!contentType.includes('markdown')) {
                throw new Error("Content is wrong content-type:\"".concat(contentType, "\" content: ").concat(escapeHtml(data.substring(0, 100))));
            }
        }
    }
    // 兼容几种常见写法
    if (data.hasOwnProperty('errorCode')) {
        // 阿里 Java 规范
        data.status = data.errorCode;
        data.msg = data.errorMessage || data.errorMsg;
    }
    else if (data.hasOwnProperty('errno')) {
        data.status = data.errno;
        data.msg = data.errmsg || data.errstr || data.msg;
    }
    else if (data.hasOwnProperty('no')) {
        data.status = data.no;
        data.msg = data.error || data.msg;
    }
    else if (data.hasOwnProperty('error')) {
        // Google JSON guide
        // https://google.github.io/styleguide/jsoncstyleguide.xml#error
        if (typeof data.error === 'object' && data.error.hasOwnProperty('code')) {
            data.status = data.error.code;
            data.msg = data.error.message;
        }
        else {
            data.status = data.error;
            data.msg = data.errmsg || data.msg;
        }
    }
    if (!data.hasOwnProperty('status')) {
        hasStatusField = false;
    }
    var payload = {
        ok: hasStatusField === false || data.status == 0,
        status: hasStatusField === false ? 0 : data.status,
        msg: data.msg || data.message,
        defaultMsg: data.defaultMsg,
        msgTimeout: data.msgTimeout,
        data: !data.data && !hasStatusField ? data : data.data // 兼容直接返回数据的情况
    };
    // 兼容返回 schema 的情况，用于 app 模式
    if (data && data.type) {
        payload.data = data;
    }
    if (payload.status == 422) {
        payload.errors = data.errors;
    }
    debug('api', 'response', payload);
    if (api.responseData && (payload.ok || !isEmpty(payload.data))) {
        debug('api', 'before dataMapping', payload.data);
        var responseData = dataMapping(api.responseData, createObject({ api: api }, normalizeApiResponseData(payload.data)), undefined, api.convertKeyToPath);
        debug('api', 'after dataMapping', responseData);
        payload.data = responseData;
    }
    return payload;
}
function wrapFetcher(fn, tracker) {
    // 避免重复处理
    if (fn._wrappedFetcher) {
        return fn;
    }
    var wrappedFetcher = function (api, data, options) {
        var _a, _b, _c;
        return __awaiter(this, void 0, void 0, function () {
            var apiCache, timeStamp;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        api = buildApi(api, data, options);
                        api.context = data;
                        if (!api.requestAdaptor) return [3 /*break*/, 2];
                        debug('api', 'before requestAdaptor', api);
                        return [4 /*yield*/, api.requestAdaptor(api, data)];
                    case 1:
                        api = (_d.sent()) || api;
                        debug('api', 'after requestAdaptor', api);
                        _d.label = 2;
                    case 2:
                        if (api.data &&
                            (api.data instanceof FormData ||
                                hasFile(api.data) ||
                                api.dataType === 'form-data')) {
                            api.data =
                                api.data instanceof FormData
                                    ? api.data
                                    : object2formData(api.data, api.qsOptions);
                        }
                        else if (api.data &&
                            typeof api.data !== 'string' &&
                            api.dataType === 'form') {
                            api.data = qsstringify(api.data, api.qsOptions);
                            api.headers = api.headers || (api.headers = {});
                            api.headers['Content-Type'] = 'application/x-www-form-urlencoded';
                        }
                        else if (api.data &&
                            typeof api.data !== 'string' &&
                            api.dataType === 'json') {
                            api.data = JSON.stringify(api.data);
                            api.headers = api.headers || (api.headers = {});
                            api.headers['Content-Type'] = 'application/json';
                        }
                        // 如果发送适配器中设置了 mockResponse
                        // 则直接跳过请求发送
                        if (api.mockResponse) {
                            return [2 /*return*/, wrapAdaptor(Promise.resolve(api.mockResponse), api, data)];
                        }
                        if (!isValidApi(api.url)) {
                            throw new Error("invalid api url:".concat(api.url));
                        }
                        debug('api', 'request api', api);
                        tracker === null || tracker === void 0 ? void 0 : tracker({ eventType: 'api', eventData: omit(api, ['config', 'data', 'body']) }, api.data);
                        if (((_a = api.method) === null || _a === void 0 ? void 0 : _a.toLocaleLowerCase()) === 'jsonp') {
                            return [2 /*return*/, wrapAdaptor(jsonpFetcher(api), api, data)];
                        }
                        if (((_b = api.method) === null || _b === void 0 ? void 0 : _b.toLocaleLowerCase()) === 'js') {
                            return [2 /*return*/, wrapAdaptor(jsFetcher(fn, api), api, data)];
                        }
                        if (typeof api.cache === 'number' && api.cache > 0) {
                            apiCache = getApiCache(api);
                            return [2 /*return*/, wrapAdaptor(apiCache
                                    ? apiCache.cachedPromise
                                    : setApiCache(api, fn(api)), api, data)];
                        }
                        // IE 下 get 请求会被缓存，所以自动加个时间戳
                        if (isIE && api && ((_c = api.method) === null || _c === void 0 ? void 0 : _c.toLocaleLowerCase()) === 'get') {
                            timeStamp = "_t=".concat(Date.now());
                            if (api.url.indexOf('?') === -1) {
                                api.url = api.url + "?".concat(timeStamp);
                            }
                            else {
                                api.url = api.url + "&".concat(timeStamp);
                            }
                        }
                        return [2 /*return*/, wrapAdaptor(fn(api), api, data)];
                }
            });
        });
    };
    wrappedFetcher._wrappedFetcher = true;
    return wrappedFetcher;
}
function wrapAdaptor(promise, api, context) {
    var _this = this;
    var adaptor = api.adaptor;
    return adaptor
        ? promise
            .then(function (response) { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        debug('api', 'before adaptor data', response.data);
                        result = adaptor(response.data, response, api, context);
                        if (!(result === null || result === void 0 ? void 0 : result.then)) return [3 /*break*/, 2];
                        return [4 /*yield*/, result];
                    case 1:
                        result = _a.sent();
                        _a.label = 2;
                    case 2:
                        debug('api', 'after adaptor data', result);
                        return [2 /*return*/, __assign(__assign({}, response), { data: result })];
                }
            });
        }); })
            .then(function (ret) { return responseAdaptor(ret, api); })
        : promise.then(function (ret) { return responseAdaptor(ret, api); });
}
/**
 * 请求远程 js 文件然后 new Function 执行，用于 schemaApi 支持 JavaScript
 * @param api
 * @returns
 */
function jsFetcher(fetcher, api) {
    return new Promise(function (resolve, reject) {
        // 大概也不会有人用 post
        api.method = 'get';
        fetcher(api).then(function (response) {
            if (typeof response.data === 'string') {
                var result = new Function('api', response.data)(api);
                resolve({
                    status: 200,
                    headers: {},
                    data: {
                        status: 0,
                        msg: '',
                        data: result
                    }
                });
            }
            else {
                reject('must return string: ' + response.data);
            }
        });
    });
}
function jsonpFetcher(api) {
    return new Promise(function (resolve, reject) {
        var _a;
        var script = document.createElement('script');
        var src = api.url;
        script.async = true;
        function remove() {
            if (script) {
                // @ts-ignore
                script.onload = script.onreadystatechange = script.onerror = null;
                if (script.parentNode) {
                    script.parentNode.removeChild(script);
                }
                script = null;
            }
        }
        var jsonp = ((_a = api.query) === null || _a === void 0 ? void 0 : _a.callback) || 'axiosJsonpCallback' + uuid();
        var old = window[jsonp];
        window[jsonp] = function (responseData) {
            window[jsonp] = old;
            var response = {
                data: responseData,
                status: 200,
                headers: {}
            };
            resolve(response);
        };
        var additionalParams = {
            _: new Date().getTime(),
            _callback: jsonp
        };
        src += (src.indexOf('?') >= 0 ? '&' : '?') + qsstringify(additionalParams);
        // @ts-ignore IE 为script.onreadystatechange
        script.onload = script.onreadystatechange = function () {
            // @ts-ignore
            if (!script.readyState || /loaded|complete/.test(script.readyState)) {
                remove();
            }
        };
        script.onerror = function () {
            remove();
            var errResponse = {
                status: 0,
                headers: {}
            };
            reject(errResponse);
        };
        script.src = src;
        document.head.appendChild(script);
    });
}
// 避免在 isApiOutdated 中修改，减少影响
function isApiOutdatedWithData(prevApi, nextApi, prevData, nextData) {
    if (!nextApi) {
        return false;
    }
    else if (!prevApi) {
        return true;
    }
    return isObjectShallowModified(buildApi(normalizeApi(prevApi), prevData), buildApi(normalizeApi(nextApi), nextData));
}
function isApiOutdated(prevApi, nextApi, prevData, nextData) {
    var _a;
    if (!nextApi) {
        return false;
    }
    // 通常是编辑器里加了属性，一开始没值，后来有了
    if (prevApi === undefined && !nextApi !== undefined) {
        return true;
    }
    nextApi = normalizeApi(nextApi);
    prevApi = (prevApi ? normalizeApi(prevApi) : prevApi);
    if (nextApi.autoRefresh === false) {
        return false;
    }
    // api 本身有变化
    if ((prevApi && prevApi.url !== nextApi.url) || !prevApi) {
        return !!(isValidApi(nextApi.url) &&
            (!nextApi.sendOn || evalExpression(nextApi.sendOn, nextData)));
    }
    var trackExpression = (_a = nextApi.trackExpression) !== null && _a !== void 0 ? _a : nextApi.url;
    if (typeof trackExpression !== 'string' || !~trackExpression.indexOf('$')) {
        return false;
    }
    var isModified = false;
    if (nextApi.trackExpression || prevApi.trackExpression) {
        isModified =
            tokenize(prevApi.trackExpression || '', prevData) !==
                tokenize(nextApi.trackExpression || '', nextData);
    }
    else {
        prevApi = buildApi(prevApi, prevData, {
            ignoreData: true
        });
        nextApi = buildApi(nextApi, nextData, {
            ignoreData: true
        });
        isModified = prevApi.url !== nextApi.url;
    }
    return !!(isModified &&
        isValidApi(nextApi.url) &&
        (!nextApi.sendOn || evalExpression(nextApi.sendOn, nextData)));
}
function isValidApi(api) {
    if (!api || typeof api !== 'string') {
        return false;
    }
    var idx = api.indexOf('://');
    // 不允许 :// 结尾
    if (~idx && idx + 3 === api.length) {
        return false;
    }
    try {
        // 不补一个协议，URL 判断为 false
        api = (~idx ? '' : "schema://domain".concat(api[0] === '/' ? '' : '/')) + api;
        new URL(api);
    }
    catch (error) {
        return false;
    }
    return true;
}
function isEffectiveApi(api, data, initFetch, initFetchOn) {
    if (!api) {
        return false;
    }
    if (initFetch === false) {
        return false;
    }
    if (initFetchOn && data && !evalExpression(initFetchOn, data)) {
        return false;
    }
    if (typeof api === 'string' && api.length) {
        return true;
    }
    else if (isObject(api) && api.url) {
        if (api.sendOn &&
            data &&
            !evalExpression(api.sendOn, data)) {
            return false;
        }
        return true;
    }
    return false;
}
function isSameApi(apiA, apiB) {
    return (apiA.method === apiB.method &&
        apiA.url === apiB.url &&
        !isObjectShallowModified(apiA.data, apiB.data, false));
}
function getApiCache(api) {
    // 清理过期cache
    var now = Date.now();
    var result;
    for (var idx = 0, len = apiCaches.length; idx < len; idx++) {
        var apiCache = apiCaches[idx];
        if (now - apiCache.requestTime > apiCache.cache) {
            apiCaches.splice(idx, 1);
            len--;
            idx--;
            continue;
        }
        if (isSameApi(api, apiCache)) {
            result = apiCache;
            break;
        }
    }
    return result;
}
function setApiCache(api, promise) {
    apiCaches.push(__assign(__assign({}, api), { cachedPromise: promise, requestTime: Date.now() }));
    return promise;
}
function clearApiCache() {
    apiCaches.splice(0, apiCaches.length);
}
function normalizeApiResponseData(data) {
    var _a;
    if (typeof data === 'undefined') {
        data = {};
    }
    else if (!isPlainObject(data)) {
        data = (_a = {},
            _a[Array.isArray(data) ? 'items' : 'result'] = data,
            _a);
    }
    return data;
}
// window.apiCaches = apiCaches;

export { buildApi, callStrFunction, clearApiCache, getApiCache, isApiOutdated, isApiOutdatedWithData, isEffectiveApi, isSameApi, isValidApi, jsFetcher, jsonpFetcher, normalizeApi, normalizeApiResponseData, responseAdaptor, setApiCache, str2AsyncFunction, str2function, wrapAdaptor, wrapFetcher };
