// amis-core/lib/utils/renderer-event.js
/**
 * amis-core v3.6.3
 * Copyright 2018-2023 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var Action = require('../actions/Action.js');
var object = require('./object.js');
var debounce = require('lodash/debounce');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var debounce__default = /*#__PURE__*/_interopDefaultLegacy(debounce);

var rendererEventListeners = [];
// 创建渲染器事件对象
function createRendererEvent(type, context) {
    var rendererEvent = {
        context: object.extendObject({ pristineData: context.data }, context),
        type: type,
        prevented: false,
        stoped: false,
        preventDefault: function () {
            rendererEvent.prevented = true;
        },
        stopPropagation: function () {
            rendererEvent.stoped = true;
        },
        get data() {
            return rendererEvent.context.data;
        },
        get pristineData() {
            return rendererEvent.context.pristineData;
        },
        setData: function (data) {
            rendererEvent.context.data = data;
        }
    };
    return rendererEvent;
}
// 绑定事件
var bindEvent = function (renderer) {
    var e_1, _a;
    var _b, _c, _d;
    if (!renderer) {
        return undefined;
    }
    var listeners = renderer.props.$schema.onEvent;
    if (listeners) {
        var _loop_1 = function (key) {
            var listener = rendererEventListeners.find(function (item) {
                return item.renderer === renderer && item.type === key;
            });
            if (listener === null || listener === void 0 ? void 0 : listener.executing) {
                (_c = (_b = listener === null || listener === void 0 ? void 0 : listener.debounceInstance) === null || _b === void 0 ? void 0 : _b.cancel) === null || _c === void 0 ? void 0 : _c.call(_b);
                rendererEventListeners = rendererEventListeners.filter(function (item) {
                    return !(item.renderer === listener.renderer && item.type === listener.type);
                });
                listener.actions.length &&
                    rendererEventListeners.push({
                        renderer: renderer,
                        type: key,
                        debounce: listener.debounce || null,
                        track: listeners[key].track || null,
                        weight: listener.weight || 0,
                        actions: listener.actions
                    });
            }
            if (!listener && ((_d = listeners[key].actions) === null || _d === void 0 ? void 0 : _d.length)) {
                rendererEventListeners.push({
                    renderer: renderer,
                    type: key,
                    debounce: listeners[key].debounce || null,
                    track: listeners[key].track || null,
                    weight: listeners[key].weight || 0,
                    actions: listeners[key].actions
                });
            }
        };
        try {
            // 暂存
            for (var _e = tslib.__values(Object.keys(listeners)), _f = _e.next(); !_f.done; _f = _e.next()) {
                var key = _f.value;
                _loop_1(key);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_f && !_f.done && (_a = _e.return)) _a.call(_e);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return function (eventName) {
            // eventName用来避免过滤广播事件
            rendererEventListeners = rendererEventListeners.filter(function (item) {
                return item.renderer === renderer && eventName !== undefined
                    ? item.type !== eventName
                    : true;
            });
        };
    }
    return undefined;
};
// 触发事件
function dispatchEvent(e, renderer, scoped, data, broadcast) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
    return tslib.__awaiter(this, void 0, void 0, function () {
        var unbindEvent, eventName, eventConfig, rendererEvent, listeners, executedCount, checkExecuted, _loop_2, listeners_1, listeners_1_1, listener, state_1, e_2_1;
        var e_2, _l;
        var _this = this;
        return tslib.__generator(this, function (_m) {
            switch (_m.label) {
                case 0:
                    unbindEvent = null;
                    eventName = typeof e === 'string' ? e : e.type;
                    (_c = (_b = (_a = renderer === null || renderer === void 0 ? void 0 : renderer.props) === null || _a === void 0 ? void 0 : _a.env) === null || _b === void 0 ? void 0 : _b.beforeDispatchEvent) === null || _c === void 0 ? void 0 : _c.call(_b, e, renderer, scoped, data, broadcast);
                    broadcast && ((_e = (_d = renderer.props).onBroadcast) === null || _e === void 0 ? void 0 : _e.call(_d, e, broadcast, data));
                    if (!broadcast) {
                        eventConfig = (_g = (_f = renderer === null || renderer === void 0 ? void 0 : renderer.props) === null || _f === void 0 ? void 0 : _f.onEvent) === null || _g === void 0 ? void 0 : _g[eventName];
                        if (!eventConfig) {
                            // 没命中也没关系
                            return [2 /*return*/, Promise.resolve()];
                        }
                        unbindEvent = bindEvent(renderer);
                    }
                    // 没有可处理的监听
                    if (!rendererEventListeners.length) {
                        return [2 /*return*/, Promise.resolve()];
                    }
                    rendererEvent = broadcast ||
                        createRendererEvent(eventName, {
                            env: (_h = renderer === null || renderer === void 0 ? void 0 : renderer.props) === null || _h === void 0 ? void 0 : _h.env,
                            nativeEvent: e,
                            data: data,
                            scoped: scoped
                        });
                    listeners = rendererEventListeners
                        .filter(function (item) {
                        return item.type === eventName &&
                            (broadcast ? true : item.renderer === renderer);
                    })
                        .sort(function (prev, next) {
                        return next.weight - prev.weight;
                    });
                    executedCount = 0;
                    checkExecuted = function () {
                        executedCount++;
                        if (executedCount === listeners.length) {
                            unbindEvent === null || unbindEvent === void 0 ? void 0 : unbindEvent(eventName);
                        }
                    };
                    _loop_2 = function (listener) {
                        var _o, _p, wait, _q, trailing, _r, leading, _s, maxWait, debounced_1, _t, trackId, trackName;
                        return tslib.__generator(this, function (_u) {
                            switch (_u.label) {
                                case 0:
                                    _o = (listener === null || listener === void 0 ? void 0 : listener.debounce) || {}, _p = _o.wait, wait = _p === void 0 ? 100 : _p, _q = _o.trailing, trailing = _q === void 0 ? true : _q, _r = _o.leading, leading = _r === void 0 ? false : _r, _s = _o.maxWait, maxWait = _s === void 0 ? 10000 : _s;
                                    if (!(listener === null || listener === void 0 ? void 0 : listener.debounce)) return [3 /*break*/, 1];
                                    debounced_1 = debounce__default["default"](function () { return tslib.__awaiter(_this, void 0, void 0, function () {
                                        return tslib.__generator(this, function (_a) {
                                            switch (_a.label) {
                                                case 0: return [4 /*yield*/, Action.runActions(listener.actions, listener.renderer, rendererEvent)];
                                                case 1:
                                                    _a.sent();
                                                    checkExecuted();
                                                    return [2 /*return*/];
                                            }
                                        });
                                    }); }, wait, {
                                        trailing: trailing,
                                        leading: leading,
                                        maxWait: maxWait
                                    });
                                    rendererEventListeners.forEach(function (item) {
                                        // 找到事件队列中正在执行的事件加上标识，下次待执行队列就会把这个事件过滤掉
                                        if (item.renderer === listener.renderer &&
                                            listener.type === item.type) {
                                            item.executing = true;
                                            item.debounceInstance = debounced_1;
                                        }
                                    });
                                    debounced_1();
                                    return [3 /*break*/, 3];
                                case 1: return [4 /*yield*/, Action.runActions(listener.actions, listener.renderer, rendererEvent)];
                                case 2:
                                    _u.sent();
                                    checkExecuted();
                                    _u.label = 3;
                                case 3:
                                    if (listener === null || listener === void 0 ? void 0 : listener.track) {
                                        _t = listener.track, trackId = _t.id, trackName = _t.name;
                                        (_k = (_j = renderer === null || renderer === void 0 ? void 0 : renderer.props) === null || _j === void 0 ? void 0 : _j.env) === null || _k === void 0 ? void 0 : _k.tracker({
                                            eventType: listener.type,
                                            eventData: {
                                                trackId: trackId,
                                                trackName: trackName
                                            }
                                        });
                                    }
                                    // 停止后续监听器执行
                                    if (rendererEvent.stoped) {
                                        return [2 /*return*/, "break"];
                                    }
                                    return [2 /*return*/];
                            }
                        });
                    };
                    _m.label = 1;
                case 1:
                    _m.trys.push([1, 6, 7, 8]);
                    listeners_1 = tslib.__values(listeners), listeners_1_1 = listeners_1.next();
                    _m.label = 2;
                case 2:
                    if (!!listeners_1_1.done) return [3 /*break*/, 5];
                    listener = listeners_1_1.value;
                    return [5 /*yield**/, _loop_2(listener)];
                case 3:
                    state_1 = _m.sent();
                    if (state_1 === "break")
                        return [3 /*break*/, 5];
                    _m.label = 4;
                case 4:
                    listeners_1_1 = listeners_1.next();
                    return [3 /*break*/, 2];
                case 5: return [3 /*break*/, 8];
                case 6:
                    e_2_1 = _m.sent();
                    e_2 = { error: e_2_1 };
                    return [3 /*break*/, 8];
                case 7:
                    try {
                        if (listeners_1_1 && !listeners_1_1.done && (_l = listeners_1.return)) _l.call(listeners_1);
                    }
                    finally { if (e_2) throw e_2.error; }
                    return [7 /*endfinally*/];
                case 8: return [2 /*return*/, Promise.resolve(rendererEvent)];
            }
        });
    });
}
var getRendererEventListeners = function () {
    return rendererEventListeners;
};
/**
 * 兼容历史配置，追加对应name的值
 * @param props
 * @param data
 * @param valueKey
 */
var resolveEventData = function (props, data, valueKey) {
    var _a, _b;
    if (valueKey === void 0) { valueKey = 'value'; }
    return object.createObject(props.data, props.name && valueKey
        ? tslib.__assign(tslib.__assign({}, data), (_a = {}, _a[props.name] = data[valueKey], _a.__rendererData = tslib.__assign(tslib.__assign({}, props.data), (_b = {}, _b[props.name] = data[valueKey], _b)), _a)) : data);
};

exports.bindEvent = bindEvent;
exports.createRendererEvent = createRendererEvent;
exports.dispatchEvent = dispatchEvent;
exports.getRendererEventListeners = getRendererEventListeners;
exports.resolveEventData = resolveEventData;
